compiler/rustc_type_ir/src/outlives.rs

```diff
@@ -56,215 +55,170 @@ pub enum Component<I: Interner> {
-        // we simply fallback to the most restrictive rule, which
-        // requires that `Pi: 'a` for all `i`.
-        ty::Alias(_, alias_ty) => {
-            if !alias_ty.has_escaping_bound_vars() {
-                // best case: no escaping regions, so push the
-                // projection and skip the subtree (thus generating no
-                // constraints for Pi). This defers the choice between
-                // the rules OutlivesProjectionEnv,
-                // OutlivesProjectionTraitDef, and
-                // OutlivesProjectionComponents to regionck.
-                out.push(Component::Alias(alias_ty));
-            } else {
-                // fallback case: hard code
-                // OutlivesProjectionComponents. Continue walking
-                // through and constrain Pi.
-                let mut subcomponents = smallvec![];
-                let mut subvisited = SsoHashSet::new();
-                compute_alias_components_recursive(tcx, ty, &mut subcomponents, &mut subvisited);
-                out.push(Component::EscapingAlias(subcomponents.into_iter().collect()));
+            ty::Placeholder(p) => {
+                self.out.push(Component::Placeholder(p));
             }
-        }
 
-        // We assume that inference variables are fully resolved.
-        // So, if we encounter an inference variable, just record
-        // the unresolved variable as a component.
-        ty::Infer(infer_ty) => {
-            out.push(Component::UnresolvedInferenceVariable(infer_ty));
-        }
+            // For projections, we prefer to generate an obligation like
+            // `<P0 as Trait<P1...Pn>>::Foo: 'a`, because this gives the
+            // regionck more ways to prove that it holds. However,
+            // regionck is not (at least currently) prepared to deal with
+            // higher-ranked regions that may appear in the
+            // trait-ref. Therefore, if we see any higher-ranked regions,
+            // we simply fallback to the most restrictive rule, which
+            // requires that `Pi: 'a` for all `i`.
+            ty::Alias(_, alias_ty) => {
+                if !alias_ty.has_escaping_bound_vars() {
+                    // best case: no escaping regions, so push the
+                    // projection and skip the subtree (thus generating no
+                    // constraints for Pi). This defers the choice between
+                    // the rules OutlivesProjectionEnv,
+                    // OutlivesProjectionTraitDef, and
+                    // OutlivesProjectionComponents to regionck.
+                    self.out.push(Component::Alias(alias_ty));
+                } else {
+                    // fallback case: hard code
+                    // OutlivesProjectionComponents. Continue walking
+                    // through and constrain Pi.
+                    let mut subcomponents = smallvec![];
+                    compute_alias_components_recursive(self.tcx, ty, &mut subcomponents);
```

Comment:
can you inline this function? It's very subtle that it uses a separate visitor/cache, so I would like to make this very explicit

Reply:
It's being called in two places -- here, and in `rustc_infer`. That would require us to expose the outlives visitor there to replicate this behavior, or just inline this one callsite? that seems weird imo